contracts/golos.worker/main.cpp

```diff
@@ -0,0 +1,1088 @@
+  {
+    LOG("proposal_id: %, comment: %, final: %", proposal_id, comment.text.c_str(), (int) finished);
+    auto proposal_ptr = get_proposal(proposal_id);
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_WORK, "invalid proposal state");
+    eosio_assert(proposal_ptr->type == proposal_t::TYPE_1, "unsupported action");
+    require_auth(proposal_ptr->worker);
+
+    get_proposals().modify(proposal_ptr, proposal_ptr->worker, [&](auto &proposal) {
+      proposal.work_status.add(comment_id, proposal.worker, comment);
+
+      if (finished)
+      {
+        proposal.set_state(proposal_t::STATE_TSPEC_AUTHOR_REVIEW);
+      }
+    });
+  }
+
+  /**
+   * @brief acceptwork accepts a work that was done by the worker. Can be called only by the technical specification author
+   * @param proposal_id proposal ID
+   * @param comment_id comment ID
+   * @param comment
+   */
+  /// @abi action
+  void acceptwork(proposal_id_t proposal_id, comment_id_t comment_id, const comment_data_t &comment)
+  {
+    LOG("proposal_id: %, comment: %", proposal_id, comment.text.c_str());
+    auto proposal_ptr = get_proposal(proposal_id);
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_TSPEC_AUTHOR_REVIEW, "invalid proposal state");
+    eosio_assert(proposal_ptr->type == proposal_t::TYPE_1, "unsupported action");
+    require_auth(proposal_ptr->tspec_author);
+
+    get_proposals().modify(proposal_ptr, proposal_ptr->tspec_author, [&](auto &proposal) {
+      proposal.set_state(proposal_t::STATE_DELEGATES_REVIEW);
+      proposal.work_status.add(comment_id, proposal.tspec_author, comment);
+    });
+  }
+
+  /**
+   * @brief reviewwork posts delegate's review
+   * @param proposal_id proposal ID
+   * @param reviewer delegate's account name
+   * @param status 0 - reject, 1 - approve. Look at the proposal_t::review_status_t
+   * @param comment_id comemnt id
+   * @param comment comment data, live empty if it isn't required
+   */
+  /// @abi action
+  void reviewwork(proposal_id_t proposal_id, account_name reviewer, uint8_t status, comment_id_t comment_id, const comment_data_t &comment)
+  {
+    LOG("proposal_id: %, comment: %, status: %, reviewer: %", proposal_id, comment.text.c_str(), (int) status, ACCOUNT_NAME_CSTR(reviewer));
+    require_app_delegate(reviewer);
+    auto proposal_ptr = get_proposal(proposal_id);
+    require_app_delegate(reviewer);
+    get_proposals().modify(proposal_ptr, reviewer, [&](proposal_t &proposal) {
+      switch (status)
+      {
+      case proposal_t::STATUS_REJECT:
+        eosio_assert(proposal.state == proposal_t::STATE_DELEGATES_REVIEW ||
+                         proposal.state == proposal_t::STATE_WORK ||
+                         proposal.state == proposal_t::STATE_TSPEC_AUTHOR_REVIEW,
+                     "invalid state " __FILE__ ":" TOSTRING(__LINE__));
+
+        proposal.review_votes.downvote(reviewer);
```

Comment:
`downvote` method doesn't allow to change vote if upvote vas previously casted

Reply:
Fixed